### PR TITLE
Use case to check if a csv file is empty

### LINKED ISSUE #213: [REQUEST] Implement CSV file is not empty (CSV rule)
**Is your feature request related to a problem? Please describe.**
A CSV file should not be empty. This is a CSV rule implemented in Google Python validator raising an EmptyFile error.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**
Line 204 in Error support priorities https://docs.google.com/spreadsheets/d/1vqe6wq7ctqk1EhYkgtZ0_TbcQ91vccfs2daSjn20BLE/edit#gid=0

### PR DESCRIPTION
**Pull request by fabrice-v**
**Summary:**

This PR adds a new use case to check for empty csv files.
Closes #213 

**Expected behavior:** 
 - the file is completely empty --> that's an error
 - the file defines headers but no row of data
   - required file --> that's an error
   - optional file --> that's a warning

In consequence, an `EmptyFileNotice` is added to the generated report

Datasets used for test taken from https://github.com/MobilityData/gtfs-validator/pull/317#issuecomment-663042812
one_empty_gtfs_file.zip
image

mixed_empty_full_gtfs_files.zip
image

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the unit tests with `gradle test` to make sure you didn't break anything
- [ ] Format the title like "Fix #<issue_number> - <short description of fix and changes>" (for example - "Fix #1111 - Check for null value before using field")
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

**Comment by fabrice-v** [comment]
@barbeau Could you test on Windows with a modified archive containing an empty file?

**Comment by barbeau** [comment]
@fabrice-v Sure, I'll test

File: application/cli-app/src/main/java/org/mobilitydata/gtfsvalidator/Main.java [hunk 1]
```diff
@@ -81,6 +81,7 @@ public static void main(String[] args) {
 
                 // base validation + build gtfs entities
                 filenameListToProcess.forEach(filename -> {
+                    config.validateCsvNotEmptyForFile(filename).execute();
```
Comment (barbeau) [inline application/cli-app/src/main/java/org/mobilitydata/gtfsvalidator/Main.java:84]: I think this should only be an error for required GTFS files? And a warning for optional files? See https://gtfs.org/reference/static#dataset-files.

**Comment by barbeau** [comment]
So tried this on my machine with a zip file that contained only 1 empty agency.txt file:

[one_empty_gtfs_file.zip](https://github.com/MobilityData/gtfs-validator/files/4966619/one_empty_gtfs_file.zip)

...and got an NPE with the same stack trace as the NPE I see for a totally empty ZIP - see https://github.com/MobilityData/gtfs-validator/issues/318.

When trying with a zip that contained an empty agency.txt file and a normal routes.txt file, I get the same NPE too:

[mixed_empty_full_gtfs_files.zip](https://github.com/MobilityData/gtfs-validator/files/4966655/mixed_empty_full_gtfs_files.zip)

File: application/cli-app/src/main/java/org/mobilitydata/gtfsvalidator/Main.java (hunk 1 above)
Comment (fabrice-v) [reply application/cli-app/src/main/java/org/mobilitydata/gtfsvalidator/Main.java:84]: It's true I interpreted an 'empty' file as a file defining zero rows of data but potentially having headers.
But the spec says
> The first line of each file must contain field names.

So I guess the only thing that shall be checked is a totally empty file (no headers), and have it being an error for any file.

**Comment by fabrice-v** [comment]
Yeah my interpretation of empty was erroneous (I considered 0 rows but headers present). Will fix, thks!

**Comment by fabrice-v** [comment]
@barbeau thks for the datasets. They shall now be processed without error. You can review the changes through they won't be merged until the tests are fixed
Also, both meaning behind 'empty' have been implemented:

>  - the file is completely empty --> that's an error
>  - the file defines headers but no row of data
>    - required file --> that's an error
>    - optional file --> that's a warning

@lionel-nj I would need your expertise to fix the tests in the branch as I modified the behavior of `ValidateAllRequiredFilePresence`
https://github.com/MobilityData/gtfs-validator/blob/cd0b5af0a59f48584398713e09bd195de7843b0f/usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateAllRequiredFilePresence.java#L63

it seemed to make sense reading the variable name at
https://github.com/MobilityData/gtfs-validator/blob/cd0b5af0a59f48584398713e09bd195de7843b0f/application/cli-app/src/main/java/org/mobilitydata/gtfsvalidator/Main.java#L57
as well as because it fixed crashes when required files are not provided.

File: usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateAllRequiredFilePresence.java [hunk 2]
```diff
@@ -52,15 +52,14 @@ public ValidateAllRequiredFilePresence(final GtfsSpecRepository specRepo,
...
      */
     public ArrayList<String> execute() {
         if (!rawFileRepo.getFilenameAll().containsAll(specRepo.getRequiredFilenameList())) {
-
             specRepo.getRequiredFilenameList().stream()
                     .filter(requiredFile -> !rawFileRepo.getFilenameAll().contains(requiredFile))
                     .forEach(missingFile -> resultRepo.addNotice(new MissingRequiredFileNotice(missingFile)));
         }
-        return new ArrayList<>(specRepo.getRequiredFilenameList());
+        return new ArrayList<>(rawFileRepo.getFilenameAll());
```
Comment (fabrice-v) [inline usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateAllRequiredFilePresence.java:63]: We'll chat about that when back, @lionel-nj

File: adapter/exporter/src/main/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporter.java [hunk 3]
```diff
@@ -266,6 +256,28 @@ public void export(MissingRequiredFileNotice toExport) throws IOException {
                 .writeTo(streamGenerator.getStream());
     }
 
+    @Override
+    public void export(EmptyFileErrorNotice toExport) throws IOException {
```
Comment (fabrice-v) [inline adapter/exporter/src/main/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporter.java:260]: Reading it, thinking it should be factorized as the only differing line concerns `.setSeverity(`

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/EmptyFileErrorNotice.java [hunk 4]
```diff
@@ -21,12 +21,11 @@
...
-
-    public CannotConstructDataProviderNotice(String filename) {
-        super(filename, E_002,
-                "Data provider error",
-                "An error occurred while trying to access raw data for file: " + filename,
+public class EmptyFileErrorNotice extends ErrorNotice {
+    public EmptyFileErrorNotice(String filename) {
+        super(filename, E_039,
+                "Empty file",
+                "File `" + filename + "` is empty.",
```
Comment (barbeau) [inline domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/EmptyFileErrorNotice.java:28]: Same here - more descriptive message would be useful, maybe something like:

```
title='Empty file', description='File `agency.txt` does not have any headers or data.'
```

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/EmptyFileWarningNotice.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.mobilitydata.gtfsvalidator.domain.entity.notice.NoticeExporter;
+import org.mobilitydata.gtfsvalidator.domain.entity.notice.base.WarningNotice;
+
+import java.io.IOException;
+
+public class EmptyFileWarningNotice extends WarningNotice {
+    public EmptyFileWarningNotice(String filename) {
+        super(filename, W_009,
+                "Empty file",
```
Comment (barbeau) [inline domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/EmptyFileWarningNotice.java:27]: Same here - maybe something like:

```
title='Empty optional file', description='File `shapes.txt` has headers but no data.'
```

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/MissingHeaderNotice.java [hunk 6]
```diff
@@ -25,8 +25,8 @@
 
     public MissingHeaderNotice(final String filename, final String missingHeaderName) {
         super(filename, E_001,
-                "Field name not defined in the GTFS specification",
-                "File `" + filename + " is missing required field name: `" + missingHeaderName + "`",
+                "Missing required field",
+                "File `" + filename + " is missing required field: `" + missingHeaderName + "`",
```
Comment (barbeau) [inline domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/MissingHeaderNotice.java:29]: You could remove the leading `"File " + ` as well to make this more concise - it should be obvious to anyone running the tool what *.txt is. Also note that currently the filename is missing the trailing tick, if you decide to keep as-is.

File: usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateAllRequiredFilePresence.java (hunk 2 above)
Comment (lionel-nj) [reply usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateAllRequiredFilePresence.java:63]: When fixing test, a few questions popped-up. Will discuss them during out net technical meeting @fabrice-v

Comment (lionel-nj) [reply usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateAllRequiredFilePresence.java]: Documentation has been updated for code clarity as well as the confusing variable names. See https://github.com/MobilityData/gtfs-validator/commit/abe16f2dd76a069a62b9b55c3561785db5c9758d

File: usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateCsvNotEmptyForFile.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        this.resultRepo = resultRepo;
+        this.logger = logger;
+    }
+
+    /**
+     * Use case execution method: for a file, checks if it empty
+     * If no file content at all is available or it contains headers but no row of data and is required,
+     * a {@link EmptyFileErrorNotice} is added to the {@link ValidationResultRepository} provided in the constructor.
+     * If a file contains headers and is optional,
+     * a {@link EmptyFileWarningNotice} is added to the {@link ValidationResultRepository} provided in the constructor.
```
Comment (barbeau) [inline usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateCsvNotEmptyForFile.java:66]:
```suggestion
     * If a file contains headers and is optional but no rows of data,
     * a {@link EmptyFileWarningNotice} is added to the {@link ValidationResultRepository} provided in the constructor.
```

File: RULES.md [hunk 8]
```diff
@@ -267,6 +268,16 @@ The `departure_time` must not precede the `arrival_time` in `stop_times.txt` if
 
 Calculated speed between stops is too fast (>150 kmh) 
 
+<a name="E047"/>
+
+### E047 - Csv file is empty
+
+Empty csv file found in the archive: file does not have any headers or data. The GTFS specification requires the first line of each file to contain field names. 
+This is related to [W012](#https://github.com/MobilityData/gtfs-validator/blob/master/RULES.md#W012).
+
```
Comment (barbeau) [inline RULES.md:277]:
```suggestion
Empty csv file found in the archive: file does not have any headers, or is a required file and does not have any data. The GTFS specification requires the first line of each file to contain field names and required files must have data.
This is related to [W012](#https://github.com/MobilityData/gtfs-validator/blob/master/RULES.md#W012).
```

File: usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/GenerateGtfsRequiredFilenameList.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+    public GenerateGtfsRequiredFilenameList(final GtfsSpecRepository gtfsSpecRepo) {
+        this.gtfsSpecRepo = gtfsSpecRepo;
+    }
+
+    /**
+     * Use case execution method: returns a list of filenames marked as required by the GTFS specification.
+     */
+    public List<String> execute() {
+        return gtfsSpecRepo.getRequiredFilenameList();
```
Comment (barbeau) [inline usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/GenerateGtfsRequiredFilenameList.java:37]: From https://github.com/MobilityData/gtfs-validator/pull/317#pullrequestreview-460894642:

>I noticed that I didn't get an error for a missing `calendar_dates.txt` or `calendar.txt` file. One of those needs to be included to have a valid GTFS, and the zip files I tested had neither.

I think this is still an issue? Could be tackled separately.

File: RULES.md [hunk 10]
```diff
@@ -36,6 +35,7 @@ Rules are declared in the [`Notice` module](https://github.com/MobilityData/gtfs
 | [E032](#E032) | `calendar.txt` `end_date` is before `start_date` |
 | [E033](#E033) | `route_id` not found |
 | [E035](#E035) | `agency_id` not found |
+| [E039](#E039) | Csv file is empty |
```
[the remark this answers is not in the document]
Comment (barbeau) [reply RULES.md]: Whoops, this is an old comment, ignore this.

File: usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/GenerateGtfsRequiredFilenameList.java (hunk 9 above)
Comment (lionel-nj) [reply usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/GenerateGtfsRequiredFilenameList.java:37]: Opening an issue regarding this.

Comment (lionel-nj) [reply usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/GenerateGtfsRequiredFilenameList.java:37]: https://github.com/MobilityData/gtfs-validator/issues/341

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/MissingHeaderNotice.java [hunk 11]
```diff
@@ -26,7 +26,7 @@
     public MissingHeaderNotice(final String filename, final String missingHeaderName) {
         super(filename, E_001,
                 "Missing required field",
-                "File `" + filename + " is missing required field name: `" + missingHeaderName + "`",
+                String.format("`%s` is missing required field: `%s`.", filename, missingHeaderName),
```
Comment (fabrice-v) [inline domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/MissingHeaderNotice.java:29]: Nice! We should use formatting going forward

**Pull request merged**